File: 백준/Gold/15824. 너 봄에는 캡사이신이 맛있단다/너 봄에는 캡사이신이 맛있단다.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main{
    static final long MOD = 1000000007L;
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static int[] pows;
    public static void main(String[] args) throws Exception{
        int n = Integer.parseInt(br.readLine());
        long answer = 0;
        int[] arr = new int[n];
        pows = new int[n+1];
        st = new StringTokenizer(br.readLine());
        for(int i=0;i<n;i++){
            arr[i] = Integer.parseInt(st.nextToken());
        }
        Arrays.sort(arr);
        for(int i=0;i<n-1;i++){
            for(int j=i+1;j<n;j++){
                answer+=((long)(arr[j]-arr[i]))*pow(j-i-1)%MOD;
            }
        }
        System.out.print(answer%MOD);
    }
    static long pow(int k){
        if(k<2) return k+1;
        if(pows[k]!=0) return pows[k];
        long prev = pow(k/2);
        return pows[k] = (int)(k%2==0?(prev*prev)%MOD:(prev*prev*2)%MOD);
    }
}

File: 백준/Gold/15824. 너 봄에는 캡사이신이 맛있단다/너 봄에는 캡사이신이 맛있단다.py
n, answer = int(input()), 0
arr = list(map(int,input().split()))
arr.sort()
pows = [0 for _ in range(n+1)]
pows[0] = 1
pows[1] = 2

def square(k):
    if pows[k]==0:
        pows[k] = 2 * (square(k // 2)**2) % 1000000007 if k % 2 == 1 else (square(k // 2)**2)
    return pows[k]


for i in range(len(arr)-1):
    for j in range(i+1,len(arr)):
        answer +=(arr[j]-arr[i])*square(j-i-1)%1000000007
print(answer%1000000007)